browser/brave_vpn/dns/brave_vpn_dns_observer_service_win.cc

```diff
@@ -99,15 +106,65 @@ bool BraveVpnDnsObserverService::IsDNSHelperLive() {
   if (dns_helper_live_for_testing_.has_value()) {
     return dns_helper_live_for_testing_.value();
   }
-  // If BraveVpnHelper is live we should not override DNS because it will be
+  // If the service is not installed we should override DNS because it will be
   // handled by the service.
-  return brave_vpn::IsBraveVPNHelperServiceLive();
+
+  if (!IsBraveVPNHelperServiceInstalled()) {
+    return false;
+  }
+  if (!IsBraveVPNHelperServiceRunning()) {
+    content::GetUIThreadTaskRunner({})->PostDelayedTask(
+        FROM_HERE,
+        base::BindOnce(&BraveVpnDnsObserverService::OnCheckIfServiceStarted,
+                       weak_ptr_factory_.GetWeakPtr()),
+        base::Seconds(kHelperServiceStartTimeoutSec));
+    // The service can be stopped and this is valid state, not started yet,
+    // crashed once and restarting and so on.
+    return true;
+  }
+  RunServiceWatcher();
+  return IsNetworkFiltersInstalled();
 }
 
-void BraveVpnDnsObserverService::LockDNS() {
-  if (IsDNSHelperLive()) {
+void BraveVpnDnsObserverService::RunServiceWatcher() {
+  service_watcher_.reset(new ServiceWatcher());
+  if (!service_watcher_->Subscribe(
+          brave_vpn::GetVpnServiceName(), SERVICE_NOTIFY_STOPPED,
+          base::BindOnce(&BraveVpnDnsObserverService::OnServiceStopped,
+                         weak_ptr_factory_.GetWeakPtr()))) {
+    VLOG(1) << "Unable to set service watcher";
+  }
+}
+
+void BraveVpnDnsObserverService::OnServiceStopped() {
+  // Postpone check because the service can be restarted by the system due to
+  // configured failure actions.
+  content::GetUIThreadTaskRunner({})->PostDelayedTask(
+      FROM_HERE,
+      base::BindOnce(&BraveVpnDnsObserverService::OnCheckIfServiceStarted,
+                     weak_ptr_factory_.GetWeakPtr()),
+      base::Seconds(kHelperServiceStartTimeoutSec));
+}
+
+bool BraveVpnDnsObserverService::IsVPNConnected() const {
+  return connection_state_.has_value() &&
+         connection_state_.value() ==
+             brave_vpn::mojom::ConnectionState::CONNECTED;
+}
+
+void BraveVpnDnsObserverService::OnCheckIfServiceStarted() {
+  if (!IsVPNConnected()) {
+    return;
+  }
+  // Checking if the service was not launched or filters were not set.
+  if (!IsNetworkFiltersInstalled() || !IsBraveVPNHelperServiceRunning()) {
```

Comment:
As we add some `DCHECK()`s in `IsNetworkFiltersInstalled()`,  I think this could be `if (!IsNetworkFiltersInstalled()) {`?

Reply:
Or checking `!IsBraveVPNHelperServiceRunning()` first?

Reply:
it doesnt matter actually because we have only one dcheck for service installed

Reply:
got it!